Sources/MessagingPush/Type/PushNotification.swift

```diff
@@ -3,9 +3,115 @@ import Foundation
 // A data type that represents a push notification.
 // Note: This data class represents *all* push notifications, even those not sent by CIO.
 public protocol PushNotification {
+    // an id that uniquely identifies this push notification.
+    // For push notifications sent by APN or FCM, this id gets populated by them with a random value.
+    // For locally created notifications, this ID is provided by the app developer. So, the ID might be a hard-coded value and not unique between 2 pushes.
     var pushId: String { get }
-    var deliveryDate: Date { get }
-    var title: String { get }
-    var message: String { get }
-    var data: [AnyHashable: Any] { get }
+
+    // Date when the push was delivered to the device.
+    // Some subclasses of UserNotifications framework does not include a delivery date so we have to keep this optional.
+    // Depending on the request that the framework makes to our SDK will determine if we get a date or not.
+    var deliveryDate: Date? { get }
+
+    var title: String { get set }
+    var body: String { get set }
+    var data: [AnyHashable: Any] { get set }
+
+    // Attachments on a push notification are mostly used to display additional content on a push such as an image.
+    var attachments: [PushAttachment] { get set }
+}
+
+public struct PushAttachment {
+    public let identifier: String
+    public let localFileUrl: URL
+}
+
+// Set of convenient properties for when a push was sent by Customer.io
+public extension PushNotification {
+    var cioAttachmentsPrefix: String {
+        "cio_sdk_"
+    }
+
+    var cioDelivery: (id: String, token: String)? {
+        guard let id = data["CIO-Delivery-ID"] as? String,
+              let token = data["CIO-Delivery-Token"] as? String
+        else {
+            return nil
+        }
+
+        return (id: id, token: token)
+    }
+
+    var isPushSentFromCio: Bool {
+        cioDelivery != nil
+    }
+
+    var cioPayload: [AnyHashable: Any]? {
+        get {
+            data["CIO"] as? [AnyHashable: Any]
+        }
+        set {
+            data["CIO"] = newValue
+        }
+    }
+
+    var cioPushPayload: [AnyHashable: Any]? {
+        get {
+            cioPayload?["push"] as? [AnyHashable: Any]
```

Comment:
is there a way to make it more type-safe? not just here but below too?

Reply:
I believe you're suggestion is in regards to the return value. Looking for something like this: 

`var cioPushPayload: CioPushPayload?` instead of the generic `[AnyHashable: Any]?`? 

I did consider that, but I do think that it makes this class more complex and harder to read without adding much value. 

I think type-safety would be really valuable for `cioPayload` and `cioPushPayload` properties, but only if they were `public` and meant to be used outside of this class. Which, these 2 properties are not designed to be used outside of this class. 

As an alternative solution, if `cioPayload` and `cioPushPayload` were set to `private`, I no longer think type-safety would be needed. 

Does that sound good? If not, have a suggestion?

Reply:
That does sound reasonable, additionally, I have not tried it but could we initialize these values on `init`? as in instead of 
first,  `cioPayload used as [AnyHashable: Any]?` and then `cioPushPayload`
retrieved via `cioPayload?["push"] as? [AnyHashable: Any]`

could we just use `data` and initialize all these variables?

Reply:
> could we initialize these values on init? 
Doing this might not be possible because `PushNotification` is a protocol and not a struct/class. We would need to do a refactor to do this idea. 

Also, I am not sure if I see value in doing this. I like how `data` is the payload's single-source-of-truth. If we were to make `cioPayload`, for example, it's own `var` variable on the `PushNotification` object instead of this `getter/setter` dynamic property, then we would need to maintain multiple variables instead of 1, `data`.

Reply:
I made the fields `private` https://github.com/customerio/customerio-ios/pull/492/commits/406143aca7caeddb10902fe219dbb9abe1aa9726